File: generic_cpp/src/generic_subscriber.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp/serialization.hpp"
#include "rclcpp/generic_subscription.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class GenericSubscriber : public rclcpp::Node
{
  public:
    GenericSubscriber(std::string topic_name) : Node("generic_subscriber"), topic_name_(topic_name)
    {
      timer_ = this->create_wall_timer(5s, std::bind(&GenericSubscriber::attempt_sub_all_msg_types, this));

      RCLCPP_INFO(this->get_logger(), "Generic Subscriber Initialized on topic: '%s'", topic_name_.c_str());
    }

  private:
  // [Fix] change function name to ...
    void attempt_sub_all_msg_types()
    {
      // get all topics and types map
      auto topics_details = this->get_topic_names_and_types();

      // search throuh to get topic name equals the specified topic
      auto topic_details = topics_details.find(topic_name_);

      // [FIX]: since we are only subscribing to one at a time because of restriction, remove the loop
      // and only subscribe to the first and most likely only type on that topic
      if (topic_details != topics_details.end() && !topic_details->second.empty()) {
        for (const auto &topic_type : topic_details->second) {
          // loop through all the message types on that topic name and subscibe to them
          RCLCPP_INFO(this->get_logger(), "Subscribing to topic '%s' with type: %s ", topic_name_.c_str(), topic_type.c_str());

          auto sub = this->create_generic_subscription(
            topic_name_,
            topic_type,
            rclcpp::QoS(10),
            std::bind(&GenericSubscriber::sub_cb, this, std::placeholders::_1, topic_type)
          );

          subscriptions_.push_back(sub);
        }

        timer_->cancel();
      } else {
        RCLCPP_INFO(this->get_logger(), "Could not find any publisher on topic: %s, retrying in 5s", topic_name_.c_str());
      }
    }

    void sub_cb(const std::shared_ptr<rclcpp::SerializedMessage> msg, std::string msg_type)
    {
      (void)msg;
      RCLCPP_INFO(this->get_logger(), "Received a message from topic '%s' with type: %s", topic_name_.c_str(), msg_type.c_str());
    }

    std::string topic_name_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::vector<rclcpp::GenericSubscription::SharedPtr> subscriptions_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  auto node = std::make_shared<GenericSubscriber>("/buffer");

  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: generic_py/generic_py/generic_subscriber.py
# from std_msgs.msg import String

import rclpy
from rclpy.node import Node
from rosidl_runtime_py.utilities import get_message

class GenericSubscriber(Node):
    def __init__(self, topic_name: str):
        super().__init__("generic_subscriber")

        self.topic_name_ = topic_name

        self.timer_ = self.create_timer(5, self.attempt_sub_to_topic)

        self.get_logger().info(f"Generic Subscriber Initialized on topic: '{topic_name}'")
    
    def attempt_sub_to_topic(self):
        topic_publishers_info = self.get_publishers_info_by_topic(self.topic_name_)

        if len(topic_publishers_info) != 0:
            self.timer_.cancel()
            self.timer_ = None

            # Only subscribe to the first one since ROS2 can only subscribe to a
            # single topic-type pair, can't sub to different types on same topic
            first_topic_info = topic_publishers_info[0]
            msg_type_str = first_topic_info.topic_type

            MsgType = get_message(msg_type_str)
            if MsgType is None:
                self.get_logger().error(f"Failed to import message type for '{msg_type_str}'")
                return
            
            self.get_logger().info(f"Subscribing to topic '{self.topic_name_}' with type: {msg_type_str}")
            self.subscription_ = self.create_subscription(MsgType, self.topic_name_, lambda msg : self.sub_cb(msg, msg_type_str), 10)
    
    def sub_cb(self, msg, msg_type):
      self.get_logger().info(f"Received a message from topic '{self.topic_name_}' with type: {msg_type}")

def main(args=None):
    rclpy.init(args=args)

    generic_sub = GenericSubscriber("/buffer")

    rclpy.spin(generic_sub)
    
    generic_sub.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
